{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hw1_answers.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyP5+P4MlGFHe3rjqpzK6nb0"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Question 1\n",
        "\n",
        "In lecture 4, we saw a table describing the numbers of SARS-CoV-2 sequencing experiments in the SRA, broken down by Library Strategy and Platform. That table was created in January 2021 from metadata describing ~190k relevant experiments that existed in the SRA at that time.\n",
        "\n",
        "This year, we have been using a table describing ~2.5 million SARS-CoV-2 experiments in the SRA. Build an equivalent experiment counts table (i.e., broken down by Library Strategy and Platform) from the 2022 SRA metadata that we have been using (https://github.com/shaunmahony/BMMB554-2022/raw/main/data/sra_ncov_bmmb554_2022.csv.gz).\n",
        "\n",
        "Hint: review the Pandas2 workbook if you’re stuck.\n",
        "\n",
        "**Do you notice any major differences in sequencing trends between this year and last year? **"
      ],
      "metadata": {
        "id": "vCmGG1p-og1I"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd"
      ],
      "metadata": {
        "id": "rC3_Trr6ogcA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "jGZE6aBznNjX",
        "outputId": "1b2df8bb-fec6-4e73-c20c-74298a94c348"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-11c4aa5e-9201-4ac5-8dd8-2031ffb01147\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Run</th>\n",
              "      <th>ReleaseDate</th>\n",
              "      <th>size_MB</th>\n",
              "      <th>Experiment</th>\n",
              "      <th>LibraryStrategy</th>\n",
              "      <th>LibrarySelection</th>\n",
              "      <th>LibrarySource</th>\n",
              "      <th>LibraryLayout</th>\n",
              "      <th>Platform</th>\n",
              "      <th>Model</th>\n",
              "      <th>SRAStudy</th>\n",
              "      <th>BioProject</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SRR17572881</td>\n",
              "      <td>2022-01-12 04:54:04</td>\n",
              "      <td>15</td>\n",
              "      <td>SRX13742219</td>\n",
              "      <td>AMPLICON</td>\n",
              "      <td>RT-PCR</td>\n",
              "      <td>VIRAL RNA</td>\n",
              "      <td>SINGLE</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>SRP333867</td>\n",
              "      <td>PRJNA750736</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SRR17572889</td>\n",
              "      <td>2022-01-12 04:54:04</td>\n",
              "      <td>21</td>\n",
              "      <td>SRX13742211</td>\n",
              "      <td>AMPLICON</td>\n",
              "      <td>RT-PCR</td>\n",
              "      <td>VIRAL RNA</td>\n",
              "      <td>SINGLE</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>SRP333867</td>\n",
              "      <td>PRJNA750736</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SRR17572888</td>\n",
              "      <td>2022-01-12 04:54:04</td>\n",
              "      <td>22</td>\n",
              "      <td>SRX13742212</td>\n",
              "      <td>AMPLICON</td>\n",
              "      <td>RT-PCR</td>\n",
              "      <td>VIRAL RNA</td>\n",
              "      <td>SINGLE</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>SRP333867</td>\n",
              "      <td>PRJNA750736</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SRR17572887</td>\n",
              "      <td>2022-01-12 04:54:04</td>\n",
              "      <td>20</td>\n",
              "      <td>SRX13742213</td>\n",
              "      <td>AMPLICON</td>\n",
              "      <td>RT-PCR</td>\n",
              "      <td>VIRAL RNA</td>\n",
              "      <td>SINGLE</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>SRP333867</td>\n",
              "      <td>PRJNA750736</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SRR17572840</td>\n",
              "      <td>2022-01-12 04:54:04</td>\n",
              "      <td>13</td>\n",
              "      <td>SRX13742260</td>\n",
              "      <td>AMPLICON</td>\n",
              "      <td>RT-PCR</td>\n",
              "      <td>VIRAL RNA</td>\n",
              "      <td>SINGLE</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>SRP333867</td>\n",
              "      <td>PRJNA750736</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-11c4aa5e-9201-4ac5-8dd8-2031ffb01147')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-11c4aa5e-9201-4ac5-8dd8-2031ffb01147 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-11c4aa5e-9201-4ac5-8dd8-2031ffb01147');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "           Run          ReleaseDate  size_MB   Experiment LibraryStrategy  \\\n",
              "0  SRR17572881  2022-01-12 04:54:04       15  SRX13742219        AMPLICON   \n",
              "1  SRR17572889  2022-01-12 04:54:04       21  SRX13742211        AMPLICON   \n",
              "2  SRR17572888  2022-01-12 04:54:04       22  SRX13742212        AMPLICON   \n",
              "3  SRR17572887  2022-01-12 04:54:04       20  SRX13742213        AMPLICON   \n",
              "4  SRR17572840  2022-01-12 04:54:04       13  SRX13742260        AMPLICON   \n",
              "\n",
              "  LibrarySelection LibrarySource LibraryLayout  Platform           Model  \\\n",
              "0           RT-PCR     VIRAL RNA        SINGLE  ILLUMINA  Illumina MiSeq   \n",
              "1           RT-PCR     VIRAL RNA        SINGLE  ILLUMINA  Illumina MiSeq   \n",
              "2           RT-PCR     VIRAL RNA        SINGLE  ILLUMINA  Illumina MiSeq   \n",
              "3           RT-PCR     VIRAL RNA        SINGLE  ILLUMINA  Illumina MiSeq   \n",
              "4           RT-PCR     VIRAL RNA        SINGLE  ILLUMINA  Illumina MiSeq   \n",
              "\n",
              "    SRAStudy   BioProject  \n",
              "0  SRP333867  PRJNA750736  \n",
              "1  SRP333867  PRJNA750736  \n",
              "2  SRP333867  PRJNA750736  \n",
              "3  SRP333867  PRJNA750736  \n",
              "4  SRP333867  PRJNA750736  "
            ]
          },
          "metadata": {},
          "execution_count": 2
        }
      ],
      "source": [
        "df = pd.read_csv('https://github.com/shaunmahony/BMMB554-2022/raw/main/data/sra_ncov_bmmb554_2022.csv.gz')\n",
        "df.head()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df.groupby(['LibraryStrategy', 'Platform']).describe()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 516
        },
        "id": "WpoHD0A-obhp",
        "outputId": "fbe2b2cb-c547-495b-b8f3-f8ba1a0e93b8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-9ddeb887-9625-4fe6-a897-008d59296a87\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead tr th {\n",
              "        text-align: left;\n",
              "    }\n",
              "\n",
              "    .dataframe thead tr:last-of-type th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th colspan=\"8\" halign=\"left\">size_MB</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "      <th>mean</th>\n",
              "      <th>std</th>\n",
              "      <th>min</th>\n",
              "      <th>25%</th>\n",
              "      <th>50%</th>\n",
              "      <th>75%</th>\n",
              "      <th>max</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>LibraryStrategy</th>\n",
              "      <th>Platform</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"4\" valign=\"top\">AMPLICON</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>2216812.0</td>\n",
              "      <td>69.277067</td>\n",
              "      <td>104.809025</td>\n",
              "      <td>0.0</td>\n",
              "      <td>15.00</td>\n",
              "      <td>46.0</td>\n",
              "      <td>82.00</td>\n",
              "      <td>21316.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ION_TORRENT</th>\n",
              "      <td>40079.0</td>\n",
              "      <td>108.539135</td>\n",
              "      <td>113.212427</td>\n",
              "      <td>0.0</td>\n",
              "      <td>56.00</td>\n",
              "      <td>87.0</td>\n",
              "      <td>132.00</td>\n",
              "      <td>5852.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>OXFORD_NANOPORE</th>\n",
              "      <td>176790.0</td>\n",
              "      <td>28.765049</td>\n",
              "      <td>36.760079</td>\n",
              "      <td>0.0</td>\n",
              "      <td>13.00</td>\n",
              "      <td>19.0</td>\n",
              "      <td>36.00</td>\n",
              "      <td>2581.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PACBIO_SMRT</th>\n",
              "      <td>199902.0</td>\n",
              "      <td>1.256341</td>\n",
              "      <td>3.328560</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.00</td>\n",
              "      <td>1.0</td>\n",
              "      <td>2.00</td>\n",
              "      <td>798.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>OTHER</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>1.0</td>\n",
              "      <td>322.000000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>322.0</td>\n",
              "      <td>322.00</td>\n",
              "      <td>322.0</td>\n",
              "      <td>322.00</td>\n",
              "      <td>322.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">RNA-Seq</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>2815.0</td>\n",
              "      <td>221.218117</td>\n",
              "      <td>485.621252</td>\n",
              "      <td>0.0</td>\n",
              "      <td>10.00</td>\n",
              "      <td>88.0</td>\n",
              "      <td>173.00</td>\n",
              "      <td>5213.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ION_TORRENT</th>\n",
              "      <td>11.0</td>\n",
              "      <td>108.909091</td>\n",
              "      <td>77.381464</td>\n",
              "      <td>0.0</td>\n",
              "      <td>45.00</td>\n",
              "      <td>122.0</td>\n",
              "      <td>139.00</td>\n",
              "      <td>265.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>OXFORD_NANOPORE</th>\n",
              "      <td>123.0</td>\n",
              "      <td>3.723577</td>\n",
              "      <td>14.157950</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.00</td>\n",
              "      <td>141.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Synthetic-Long-Read</th>\n",
              "      <th>PACBIO_SMRT</th>\n",
              "      <td>2.0</td>\n",
              "      <td>110.500000</td>\n",
              "      <td>3.535534</td>\n",
              "      <td>108.0</td>\n",
              "      <td>109.25</td>\n",
              "      <td>110.5</td>\n",
              "      <td>111.75</td>\n",
              "      <td>113.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Targeted-Capture</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>3565.0</td>\n",
              "      <td>39.373072</td>\n",
              "      <td>88.707368</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.00</td>\n",
              "      <td>6.0</td>\n",
              "      <td>30.00</td>\n",
              "      <td>713.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">WGA</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>20993.0</td>\n",
              "      <td>27.065546</td>\n",
              "      <td>19.223225</td>\n",
              "      <td>0.0</td>\n",
              "      <td>19.00</td>\n",
              "      <td>23.0</td>\n",
              "      <td>30.00</td>\n",
              "      <td>951.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ION_TORRENT</th>\n",
              "      <td>724.0</td>\n",
              "      <td>110.338398</td>\n",
              "      <td>84.355748</td>\n",
              "      <td>1.0</td>\n",
              "      <td>62.00</td>\n",
              "      <td>99.0</td>\n",
              "      <td>134.25</td>\n",
              "      <td>1045.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>OXFORD_NANOPORE</th>\n",
              "      <td>2910.0</td>\n",
              "      <td>61.658419</td>\n",
              "      <td>44.027455</td>\n",
              "      <td>0.0</td>\n",
              "      <td>25.00</td>\n",
              "      <td>56.0</td>\n",
              "      <td>86.00</td>\n",
              "      <td>375.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"4\" valign=\"top\">WGS</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>15650.0</td>\n",
              "      <td>53.147923</td>\n",
              "      <td>80.065259</td>\n",
              "      <td>0.0</td>\n",
              "      <td>19.00</td>\n",
              "      <td>37.0</td>\n",
              "      <td>57.00</td>\n",
              "      <td>3077.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ION_TORRENT</th>\n",
              "      <td>117.0</td>\n",
              "      <td>119.059829</td>\n",
              "      <td>96.455751</td>\n",
              "      <td>0.0</td>\n",
              "      <td>35.00</td>\n",
              "      <td>105.0</td>\n",
              "      <td>170.00</td>\n",
              "      <td>402.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>OXFORD_NANOPORE</th>\n",
              "      <td>3661.0</td>\n",
              "      <td>25.172357</td>\n",
              "      <td>98.275539</td>\n",
              "      <td>0.0</td>\n",
              "      <td>8.00</td>\n",
              "      <td>9.0</td>\n",
              "      <td>15.00</td>\n",
              "      <td>3631.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PACBIO_SMRT</th>\n",
              "      <td>48.0</td>\n",
              "      <td>29.229167</td>\n",
              "      <td>23.679814</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.75</td>\n",
              "      <td>39.5</td>\n",
              "      <td>51.00</td>\n",
              "      <td>57.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WXS</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>2.0</td>\n",
              "      <td>10.500000</td>\n",
              "      <td>2.121320</td>\n",
              "      <td>9.0</td>\n",
              "      <td>9.75</td>\n",
              "      <td>10.5</td>\n",
              "      <td>11.25</td>\n",
              "      <td>12.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>miRNA-Seq</th>\n",
              "      <th>ILLUMINA</th>\n",
              "      <td>20.0</td>\n",
              "      <td>35.050000</td>\n",
              "      <td>21.770272</td>\n",
              "      <td>1.0</td>\n",
              "      <td>22.25</td>\n",
              "      <td>30.0</td>\n",
              "      <td>51.75</td>\n",
              "      <td>78.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9ddeb887-9625-4fe6-a897-008d59296a87')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-9ddeb887-9625-4fe6-a897-008d59296a87 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-9ddeb887-9625-4fe6-a897-008d59296a87');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                       size_MB                                 \\\n",
              "                                         count        mean         std    min   \n",
              "LibraryStrategy     Platform                                                    \n",
              "AMPLICON            ILLUMINA         2216812.0   69.277067  104.809025    0.0   \n",
              "                    ION_TORRENT        40079.0  108.539135  113.212427    0.0   \n",
              "                    OXFORD_NANOPORE   176790.0   28.765049   36.760079    0.0   \n",
              "                    PACBIO_SMRT       199902.0    1.256341    3.328560    0.0   \n",
              "OTHER               ILLUMINA               1.0  322.000000         NaN  322.0   \n",
              "RNA-Seq             ILLUMINA            2815.0  221.218117  485.621252    0.0   \n",
              "                    ION_TORRENT           11.0  108.909091   77.381464    0.0   \n",
              "                    OXFORD_NANOPORE      123.0    3.723577   14.157950    0.0   \n",
              "Synthetic-Long-Read PACBIO_SMRT            2.0  110.500000    3.535534  108.0   \n",
              "Targeted-Capture    ILLUMINA            3565.0   39.373072   88.707368    0.0   \n",
              "WGA                 ILLUMINA           20993.0   27.065546   19.223225    0.0   \n",
              "                    ION_TORRENT          724.0  110.338398   84.355748    1.0   \n",
              "                    OXFORD_NANOPORE     2910.0   61.658419   44.027455    0.0   \n",
              "WGS                 ILLUMINA           15650.0   53.147923   80.065259    0.0   \n",
              "                    ION_TORRENT          117.0  119.059829   96.455751    0.0   \n",
              "                    OXFORD_NANOPORE     3661.0   25.172357   98.275539    0.0   \n",
              "                    PACBIO_SMRT           48.0   29.229167   23.679814    0.0   \n",
              "WXS                 ILLUMINA               2.0   10.500000    2.121320    9.0   \n",
              "miRNA-Seq           ILLUMINA              20.0   35.050000   21.770272    1.0   \n",
              "\n",
              "                                                                     \n",
              "                                        25%    50%     75%      max  \n",
              "LibraryStrategy     Platform                                         \n",
              "AMPLICON            ILLUMINA          15.00   46.0   82.00  21316.0  \n",
              "                    ION_TORRENT       56.00   87.0  132.00   5852.0  \n",
              "                    OXFORD_NANOPORE   13.00   19.0   36.00   2581.0  \n",
              "                    PACBIO_SMRT        0.00    1.0    2.00    798.0  \n",
              "OTHER               ILLUMINA         322.00  322.0  322.00    322.0  \n",
              "RNA-Seq             ILLUMINA          10.00   88.0  173.00   5213.0  \n",
              "                    ION_TORRENT       45.00  122.0  139.00    265.0  \n",
              "                    OXFORD_NANOPORE    0.00    0.0    1.00    141.0  \n",
              "Synthetic-Long-Read PACBIO_SMRT      109.25  110.5  111.75    113.0  \n",
              "Targeted-Capture    ILLUMINA           1.00    6.0   30.00    713.0  \n",
              "WGA                 ILLUMINA          19.00   23.0   30.00    951.0  \n",
              "                    ION_TORRENT       62.00   99.0  134.25   1045.0  \n",
              "                    OXFORD_NANOPORE   25.00   56.0   86.00    375.0  \n",
              "WGS                 ILLUMINA          19.00   37.0   57.00   3077.0  \n",
              "                    ION_TORRENT       35.00  105.0  170.00    402.0  \n",
              "                    OXFORD_NANOPORE    8.00    9.0   15.00   3631.0  \n",
              "                    PACBIO_SMRT        0.75   39.5   51.00     57.0  \n",
              "WXS                 ILLUMINA           9.75   10.5   11.25     12.0  \n",
              "miRNA-Seq           ILLUMINA          22.25   30.0   51.75     78.0  "
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Observations:\n",
        "\n",
        "The first thing to note is that the number of sequenced SARS-CoV-2 genomes is now over 2.5 million, a remarkable increase from the ~190k genomes sequenced this time last year. Illumina remains the dominant sequencing technology, with 84% of all genomes. Some of the other sequencing technologies showed strong growth; for example PacBio and Ion Torrent showed strong year-on-year application growth. Amplicon sequencing is also the dominant library construction approach.  "
      ],
      "metadata": {
        "id": "GsCj4IGgq2GG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Question 2\n",
        "\n",
        "Wordle has been taking the internet by storm over the past few weeks. For those (few) of you who have not heard of it yet, it’s a simple word guessing game. The goal is to guess a random 5-letter word. You get 6 guesses. After each guess, the letters in your guess turn green if the guessed letter is in the correct position, orange if the letter is in the answer word but you have it in the wrong position, and grey if the letter you typed is not in the answer word. So, you can gain some information from each guess that will inform your next guess. If you haven’t already, go play a game to get a sense of how it works: https://www.powerlanguage.co.uk/wordle. Note that only one game is made available per day.\n",
        "\n",
        "After playing, you’ll probably start to wonder whether there is some “best” or optimally informative word that you should use as your starting guess. This is an interesting question, and has connections to information theoretic problems that we come across in bioinformatics all the time. In this homework question, I want you to use Python to come up with some answers.\n",
        "\n",
        "Here’s a list of all 5-letter words that Wordle accepts as legitimate guesses: https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/wordlewords.txt\n",
        "\n",
        "Your job is to find the most informative word. You can define “informative” as you wish, as long as two conditions are met:\n",
        "\n",
        "You clearly describe your definition and rationale in the workbook.\n",
        "You perform a search over all words in the above file to find a word that maximizes your definition of “informative”.\n",
        "If it were me, I would stick with a classic information content approach. You could first count how often each possible letter is used in the list of words. The word with the highest information content would then be a word that contains a combination of the most frequently occurring letters. Such a word would literally be the word that maximizes your chances to turn Wordle boxes green or orange. More specifically, the approach could be:\n",
        "\n",
        "Count the occurrences of each letter (a-z) in the file. These counts become the “score” associated with each letter.\n",
        "Score each possible word by summing the per-letter scores.\n",
        "Sort the words by score.\n",
        "Provide me with a list of the 10 most informative words and the 10 least informative words."
      ],
      "metadata": {
        "id": "KwTzF2fJrxLj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/wordlewords.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NtAR5wWDrwUi",
        "outputId": "2c0c0735-0347-4ee9-eddc-48e4a7a9355c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-03-14 20:49:39--  https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/wordlewords.txt\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 77832 (76K) [text/plain]\n",
            "Saving to: ‘wordlewords.txt’\n",
            "\n",
            "\rwordlewords.txt       0%[                    ]       0  --.-KB/s               \rwordlewords.txt     100%[===================>]  76.01K  --.-KB/s    in 0.02s   \n",
            "\n",
            "2022-03-14 20:49:39 (3.58 MB/s) - ‘wordlewords.txt’ saved [77832/77832]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#From Troy\n",
        "\n",
        "letter_counts = {} # Initialize dictionary for the letter counts\n",
        "\n",
        "with open('wordlewords.txt','r') as wordfile: # Open the file of wordle words\n",
        "  for line in wordfile: # Loop through each line in file\n",
        "    word = line.rstrip() # Remove potential newline characters\n",
        "\n",
        "    for character in word: # Loop through each character in the word from that line\n",
        "      letter_counts.setdefault(character,0) # Make sure if there isn't already a dictionary key for the letter to add it and set count = 0\n",
        "      letter_counts[character] += 1 # Add to the count for each letter\n",
        "\n",
        "word_scores = {} # Initialize dictionary for the words and their scores\n",
        "\n",
        "with open('wordlewords.txt','r') as wordfile: # Open the file of wordle words\n",
        "  for line in wordfile: # Loop through each line in file\n",
        "    word = line.rstrip() # Remove potential newline characters\n",
        "    word_scores.setdefault(word,0) # Make sure if there isnt already a dictionary key for the word (there shouldn't be) to add it and set count = 0\n",
        "\n",
        "    for character in word: # Loop through each character in the word from that line\n",
        "      word_scores[word] += letter_counts[character] # Sum the counts for the word\n",
        "\n",
        "sorted_word_scores = sorted(word_scores.items(), key=lambda word:word[1]) # Use the built in sorted function and sort dict by value\n"
      ],
      "metadata": {
        "id": "JDErqcbps4BN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# From Troy\n",
        "\n",
        "lowest_scoring_words = [] # Initialize list for lowest scoring words\n",
        "\n",
        "for i in range(10): # Add the 10 lowest scoring words to the list\n",
        "  lowest_scoring_words.append(sorted_word_scores[i][0])\n",
        "\n",
        "print(\"Lowest scoring words: (Lowest scoring first)\")\n",
        "print(lowest_scoring_words)\n",
        "\n",
        "highest_scoring_words = [] # Initialize list for highest scoring words\n",
        "\n",
        "for i in range(10): # Add the 10 highest scoring words to the list\n",
        "\n",
        "  index = (len(sorted_word_scores)-(i+1)) #Index from the end of the list\n",
        "\n",
        "  highest_scoring_words.append(sorted_word_scores[index][0]) # Add the highest scoring words to the list\n",
        "\n",
        "print(\"Highest scoring words: (Highest scoring first)\")\n",
        "print(highest_scoring_words)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vph6AM73NJBu",
        "outputId": "e9a2f68e-5919-4789-85a7-102621698010"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Lowest scoring words: (Lowest scoring first)\n",
            "['fuzzy', 'buzzy', 'huzzy', 'whizz', 'muzzy', 'fizzy', 'fuffy', 'bizzy', 'jiffy', 'phizz']\n",
            "Highest scoring words: (Highest scoring first)\n",
            "['esses', 'sessa', 'sasse', 'asses', 'eases', 'sease', 'seers', 'erses', 'seres', 'reses']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## From Troy\n",
        "\n",
        "best_word_score = 0\n",
        "\n",
        "# For each word in the dictionary of words\n",
        "for entry in word_scores:\n",
        "\n",
        "  # If the score of that word is bigger than the previous best\n",
        "  if word_scores[entry] > best_word_score:\n",
        "    \n",
        "    unique_chars = [] \n",
        "    unique_letters = True\n",
        "\n",
        "    for character in entry: # Loop through each character in the word\n",
        "      if character not in unique_chars: # If the character is not already in the word: \n",
        "        unique_chars.append(character) # Add it to the list of characters in the word\n",
        "      else:\n",
        "        unique_letters = False # If the character is already in the word the word does not have all unique letters\n",
        "    \n",
        "    if unique_letters == True: # If the word has all unique letters and is higher scoring than the previous best this is the best word\n",
        "      best_word = entry\n",
        "      best_word_score = word_scores[entry]\n",
        "      print(best_word, \" \", best_word_score)\n",
        "print(\"Most Informative word: \", best_word) # Print the most informative word in the worldle word list"
      ],
      "metadata": {
        "id": "zQyeBkKr6QU5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# From Will\n",
        "\n",
        "#This part of the code defines a function use to select out those words with repeated letters in the word\n",
        "def isNotReplicated(word):\n",
        "  bucket = []\n",
        "  for i in range(5):\n",
        "    if word[i] in bucket:\n",
        "      return False\n",
        "    bucket.append(word[i])\n",
        "  return True\n",
        "\n",
        "\n",
        "#Construct list of words that contain no repeated letters\n",
        "word_list = sorted(word_scores.items(), key=lambda k: k[1]) #sort word in the list by their scores\n",
        "non_repeat_list = []\n",
        "for word,_ in word_list:\n",
        "  if isNotReplicated(word):\n",
        "    non_repeat_list.append(word)\n",
        "\n",
        "print('Most informative words: ', non_repeat_list[-10:]) "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nZAzb8jQOEH3",
        "outputId": "f0f818f7-0644-442d-dc17-3d421d206199"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Most informative words:  ['stoae', 'aloes', 'serai', 'aesir', 'raise', 'reais', 'arise', 'arose', 'aeros', 'soare']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Question 3\n",
        "\n",
        "In Lecture 5, we familiarized ourselves with the FASTQ file format in this notebook: (https://colab.research.google.com/github/shaunmahony/BMMB554-2022/blob/master/ipynb/FASTQ.ipynb). We learned how the quality string letters encode probability values, which in turn represent the probabilities that the corresponding sequenced base is incorrectly called by the sequencer. We also defined functions in the workbook that can help you to convert between FASTQ quality letters and the probability values.\n",
        "\n",
        "Here, I want you to use a collection of real Illumina FASTQ quality strings to simulate “sequencing errors” in a collection of sequences that have been sampled at random from the human genome (hg38 build). You will then map the “simulated” reads back to the genome (using your choice of aligner like BWA or Bowtie). Compare the uniquely mapping read rates between the perfectly sampled reads and the simulated sequencing error reads. Use your observations to answer the question: Do typical sequencing error rates impact read alignment?\n",
        "\n",
        "Here’s the data you’ll need:\n",
        "\n",
        "1M perfectly sampled 50bp sequences from hg38\n",
        "FASTA format (gzipped): https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/rand1M.fa.gz\n",
        "FASTQ format (gzipped) with fake “perfect” quality scores: https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/rand1M.fq.gz\n",
        "1M Illumina quality scores (50bp length)\n",
        "TXT format (gzipped): https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/quals1M.txt.gz\n",
        "How will you simulate sequencing errors? I would do something like the following:\n",
        "\n",
        "Read in the FASTA sequences and the quality score strings. Pair them up (there are 1M of each).\n",
        "Iterate through each sequence & quality string pair:\n",
        "Iterate through each character in the quality string\n",
        "Convert the quality character to a p-value\n",
        "Roll a dice with a random number generator.\n",
        "If the random number is less than the p-value, “mutate” the base in the corresponding position in the sequence. If the sequence letter was an A, replace with a random choice from C,G,T, etc.\n",
        "Otherwise leave the base in the corresponding sequence position alone.\n",
        "Print all sequence/quality string pairs out in FASTQ format (the name of the sequence can be random or related to the FASTA header - doesn’t really matter).\n",
        "After making your simulated sequences, map them and the “perfect” FASTQ format file above to hg38 (Galaxy or command line is fine). Check mapping rates.\n",
        "\n",
        "If you’re interested, you can further check whether the reads are being mapped to the correct positions (there are contained in the names of the FASTA sequences), but this is not required for the homework assignment."
      ],
      "metadata": {
        "id": "63YaE2H_tHHa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import random"
      ],
      "metadata": {
        "id": "IbcoE-RJXnqn"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "wget https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/rand1M.fa.gz\n",
        "wget https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/rand1M.fq.gz\n",
        "wget https://raw.githubusercontent.com/shaunmahony/BMMB554-2022/main/homework/hw1/quals1M.txt.gz\n",
        "gunzip rand1M.fa.gz\n",
        "gunzip rand1M.fq.gz\n",
        "gunzip quals1M.txt.gz"
      ],
      "metadata": {
        "id": "HunwqbUwtItR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def phred33_to_q(qual):\n",
        "  \"\"\" Turn Phred+33 ASCII-encoded quality into Phred-scaled integer \"\"\"\n",
        "  return ord(qual)-33\n",
        "\n",
        "def q_to_phred33(Q):\n",
        "  \"\"\" Turn Phred-scaled integer into Phred+33 ASCII-encoded quality \"\"\"\n",
        "  return chr(Q + 33)\n",
        "\n",
        "def q_to_p(Q):\n",
        "  \"\"\" Turn Phred-scaled integer into error probability \"\"\"\n",
        "  return 10.0 ** (-0.1 * Q)\n",
        "\n",
        "def p_to_q(p):\n",
        "    \"\"\" Turn error probability into Phred-scaled integer \"\"\"\n",
        "    import math\n",
        "    return int(round(-10.0 * math.log10(p))) "
      ],
      "metadata": {
        "id": "znENONjhtKJM"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Parsing the file \n",
        "# Function for parsing a FASTA file into a Python dictionary, dict maps short names to corresponding nucleotide strings (w/o whitespace)\n",
        "def parse_fasta(fh):\n",
        "    fa = {}\n",
        "    current_short_name = None #'None' is an empty string \n",
        "\n",
        "    # Part 1: compile list of lines per sequence\n",
        "    for ln in fh: #iterate over lines (ln) in fasta file\n",
        "        if ln[0] == '>':\n",
        "\n",
        "            # new name line; remember current sequence's short name\n",
        "            long_name = ln[1:].rstrip()\n",
        "            current_short_name = long_name.split()[0]\n",
        "            fa[current_short_name] = []\n",
        "        else:\n",
        "            # append nucleotides to current sequence\n",
        "            fa[current_short_name].append(ln.rstrip())\n",
        "    # Part 2: join lists into strings\n",
        "    for short_name, nuc_list in fa.items():\n",
        "        # join this sequence's lines into one long string\n",
        "        fa[short_name] = ''.join(nuc_list)\n",
        "    return fa"
      ],
      "metadata": {
        "id": "0H2F0sy-1Eb6"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "with open('rand1M.fa', 'r') as fasta_file : #Open FASTA file\n",
        "  fa = parse_fasta(fasta_file) #Parse FASTA file\n",
        "\n",
        "fa_seqs = list(fa.values())\n",
        "\n",
        "# Parsing the quality scores data\n",
        "with open(\"quals1M.txt\") as j:\n",
        "  quals1M = j.read().splitlines()\n",
        "quality = list(quals1M)"
      ],
      "metadata": {
        "id": "38CX5Sgi1FFN"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fa_seqs"
      ],
      "metadata": {
        "id": "gO3Ame8fRk5H"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# From Paulina\n",
        "\n",
        "# The reason for choosing to mutate only if less than p value, is because the p value is the probability of a wrong call..\n",
        "# we are using the same odds to mutate the base into something else\n",
        "\n",
        "\n",
        "mutatedfa = []\n",
        "\n",
        "for i in range(len(fa_seqs)):\n",
        "  original = fa_seqs[i]\n",
        "  perfect_quals = quality[i]\n",
        "  j = 0\n",
        "\n",
        "  newseq = []\n",
        "\n",
        "  for quals in perfect_quals:\n",
        "    Q = phred33_to_q(quals)\n",
        "    p = q_to_p(Q)\n",
        "    randomnums = random.random()\n",
        "\n",
        "    if randomnums < p:\n",
        "      if original[j] == \"A\":\n",
        "        mutated = random.choice(\"CGT\")\n",
        "      elif original[j] == \"T\":\n",
        "        mutated = random.choice(\"CGA\")\n",
        "      elif original[j] == \"C\":\n",
        "        mutated = random.choice(\"AGT\")\n",
        "      elif original[j] == \"G\":\n",
        "        mutated = random.choice(\"CAT\")\n",
        "    else:\n",
        "      mutated = original[j]\n",
        "    newseq.append(mutated)\n",
        "    j += 1\n",
        "  new_sequence = ''.join(newseq)\n",
        "  mutatedfa.append(new_sequence)"
      ],
      "metadata": {
        "id": "osAq_kXl1XCg"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#From Paulina (with edits)\n",
        "\n",
        "# Checking how many sequences are still identical\n",
        "k = []\n",
        "for i in range(len(mutatedfa)) :\n",
        "  if mutatedfa[i] == fa_seqs[i]:\n",
        "    k.append(i)\n",
        "\n",
        "print('Identical sequences: ', len(k)) \n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jSc_LJnRRHF_",
        "outputId": "964a6112-d8f9-4448-8668-0a353c751542"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Identical sequences:  851468\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#From Paulina (with edits) \n",
        "\n",
        "# 1. Read name = same as original (it contains the original coordinates)\n",
        "# 2. Nucleotide sequence = mutated sequence list = \"mutatedfa\"\n",
        "# 3. Plus variable is '+' sign \n",
        "# 4. Quality sequence = quality scores list = \"quality\"\n",
        "\n",
        "names = []\n",
        "for i in fa.keys() :\n",
        "    names.append(\"@\"+i)\n",
        "\n",
        "plus = '+' * len(names)\n",
        "\n",
        "merge = [i for i in zip(names,mutatedfa,plus,quality)] \n",
        "x=[]\n",
        "for i in merge:\n",
        "  for j in i:\n",
        "    x.append(j)\n",
        "x\n",
        "\n",
        "# Saving the mutated sequence\n",
        "np.savetxt('mut_rand1M.fq', x, fmt='%s')"
      ],
      "metadata": {
        "id": "Kne6pB1IVtIx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Results\n",
        "\n",
        "I mapped the \"perfect\" and \"mutated\" versions of the randomly selected reads back to hg38 using BWA-aln. \n",
        "\n",
        "**946926/1000000 = 94.69% of \"perfect\" reads map back to the genome..**\n",
        "\n",
        "...waitaminute...\n",
        "\n",
        "Why do only 94.69% of sequences that have been perfectly copied out of the hg38 genome map back to the hg38 genome? \n",
        "\n",
        "Well... did you notice the reads containing Ns in the sequences?"
      ],
      "metadata": {
        "id": "CWDqYzKLCmnb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "grep -c N mut_rand1M.fq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ok8S0i49XwwX",
        "outputId": "515b294e-2be2-4a1c-a2bc-c64f8cfd82a7"
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "53078\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "grep -c NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN mut_rand1M.fq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "67nvQ3ohqxTT",
        "outputId": "482e7faf-910b-4949-e245-6b10e410d0fe"
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "52991\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Mapping rates after removing N-containing reads\n",
        "\n",
        "Over 5% of all positions on the genome are unmappable because the reference genome is not defined at these positions. \n",
        "\n",
        "These Ns are solely responsible for the \"missing\" accuracy in BWA alignments of perfectly sampled reads. For example, if we remove all N-containing reads and map against the genome again, we find: \n",
        "\n",
        "**946922/946922 = 100% of non-N \"perfect\" reads map back to the genome**"
      ],
      "metadata": {
        "id": "JQdXutjIMdQ3"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Using the same approach (i.e., first filtering out N-containing reads), the \"mutated\" sequences map at the following rate: \n",
        "\n",
        "**911307/946922 = 96.24% of non-N \"mutated\" reads map back to the genome**\n",
        "\n",
        "From this, we can conclude the following: \n",
        "**If Illumina base-calling errors follow the error rates specified in FASTQ quality scores, 3-4% of (50bp) reads will be unmappable because of base-calling errors alone.**\n"
      ],
      "metadata": {
        "id": "RJ018VZGNUFB"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Moving away from N-containing reference genomes\n",
        "\n",
        "As an aside, the \"telomere-to-telomere\" (T2T) consortium aims to produce a complete human genome assembly that removes all the unknown \"N\" regions. You can read about their work \n",
        "[here](https://sites.google.com/ucsc.edu/t2tworkinggroup).\n",
        "\n",
        "The T2T consortium uses a cell line (CHM13) derived from a hydatidiform mole. Its genome is almost entirely homozygous, thus removing assembly issues due to heterozygous alleles. The use of long read sequencing technologies enabled complete read-through and assembly of the genome. \n",
        "\n",
        "What happens if we sample 1,000,000 \"reads\" from the CHM13 genome, and map back to hg38? There are no Ns in these reads, but the genotype is different from those sequenced during the human genome project, and we are now pulling \"reads\" from telomeric and centromeric regions that are not present in the hg38 assembly (as we would be in a real research project). \n",
        "\n",
        "**989433/1000000 = 98.94% of perfectly sampled CHM13 reads get mapped back to the hg38 genome**. \n",
        "\n",
        "We can conclude from this that a (relatively small) portion of reads would be unmappable to the reference genome based solely on differences between the cells' genotype and the reference. "
      ],
      "metadata": {
        "id": "UROXfz6tOgXR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Are the reads mapping to the correct locations?\n",
        "\n",
        "*(kudos to Polina for digging into this question)* \n",
        "\n",
        "Thus far, we've just asked whether the reads map to the genome. But recall that the reads are sampled from actual genomic positions. We should be interested in whether the reads are mapping back to their correct locations. By comparing the mapping positions in the BAM files to the actual positions in the read names, I cam up with the following statistics:\n",
        "\n",
        "**873200/946922 = 92.2% of non-N \"perfect\" reads map back to their correct locations on the genome.**\n",
        "\n",
        "Why not 100%? \n",
        "\n",
        "In the mutated reads:\n",
        "**838291/946922 = 88.5% of non-N \"mutated\" reads map back to their correct locations on the genome.**\n",
        "This is consistent with the error rate in mutated reads above, suggesting that base-calling errors are mainly making reads not map to the genome at all, rather than making them map to erroneousl positions. \n"
      ],
      "metadata": {
        "id": "YrsW23EEU46T"
      }
    }
  ]
}